pkg/virt-config/deprecation/macvtap.go

```diff
@@ -25,6 +25,10 @@ import (
 
 const MacvtapDiscontinueMessage = "Macvtap network binding is discontinued since v1.3. Please refer to Kubevirt user guide for alternatives."
 
+func init() {
```

Comment:
The current code uses kind of global var (`featureGates` slice) and init() function that initialize it every time the `deprecation` package is imported (using `init()`).
This approach makes the code challenging to follow and has more potential for side-effects (due to how `init` function acts).
I suggest to not extend the usage of it where possible (using `init()` + `RegisterFeatureGate()` at the FR file).

Reply:
> The current code uses kind of global var (`featureGates` slice) and init() function that initialize it every time the `deprecation` package is imported (using `init()`).

There is no much change from what existed so far.
Go evaluates consts, then variables and lastly `init()`.
The practical change here is that all FGs are processed through the registration function, where the data can be normalized (e.g. setting a default message) and it is not limited to be called/defined from a single location. 

> This approach makes the code challenging to follow and has more potential for side-effects (due to how `init` function acts). I suggest to not extend the usage of it where possible (using `init()` + `RegisterFeatureGate()` at the FR file).

What exactly is challenging and which side-effects are you worried about?
The practice of using `init()` is pretty common and it looks natural to be used here.

We need the FG pool to be initialized once per application (binary) and this is exactly what happens here. As we only need a single instance of it per application, we call allow the registrations to be performed before `main()` is called.

But if you see a potential problem, please be specific and share.

Reply:
The current state is no better, the fact that we use init function for initializing the featureGate list.
I think we could achieve the same with more simple solution.

With this change its now possible to cross the board and modify the featureGate function from different places through the init + registaion function, its looks hacky and somewhat of misuse of init function.

IMO its not trivial solution, using an explicit functions (and maybe complimentary) type would make it easier to follow and understand, having said that I understand it require more work and re-arranging some files.

Reply:
I am not convinced it is complicated, perhaps you can show why it is complicated by comparing scenario usages for adding a new FG to the pool.

Reply:
@EdDev I will try to rephrase the concern - it seems like the right way to use `init()` is either in the `main` package of a program, or if the package is some kind of a service like `net/http/pprof` and then it makes sense to import it only once, from the main program source file. I am not sure though what can be harmful with redundant calls of `init()` in our case, we can use `once.Do(f)` inside the `init` to make sure the code runs only once.
@ormergi The usage of `init` was there before this PR, is this necessary to improve it in this PR?

Reply:
@enp0s3 , I think it is the opposite: `init()` is mainly used in libraries, not in `main()` or `services`.
Given a library, the usage of `init()` is equivalent to defining a variable with some initial data in it. It happens once per execution and does not require an explicit call from the main flow.

A package can be imported many times, but the `init()` is still called only once, the same way a const or a variable is evaluated only once.
If two `init()` register the same FG, then the last one overrides the prev. But this is a controlled decision, we could have chosen to blow up or ignore it. This is no different from calling all this from a regular function.